### PR TITLE
Point interactivity

### PR DESCRIPTION
**Pull request by davidmanzanares**
Fix https://github.com/CartoDB/renderer-prototype/issues/198

### CODE REVIEW

File: src/core/dataframe.js [hunk 1]
```diff
@@ -65,22 +65,45 @@ export default class Dataframe {
         this.style = style;
```
Comment (Jesus89) [inline src/core/dataframe.js]: I think this `setStyle` function can be removed

Comment (davidmanzanares) [reply src/core/dataframe.js]: Yes, that's legacy code

Comment (davidmanzanares) [reply src/core/dataframe.js]: Fixed

File: src/core/dataframe.js [hunk 2]
```diff
@@ -61,26 +61,45 @@ export default class Dataframe {
...
             default:
                 return [];
         }
     }
 
-    _getPointsAtPosition(pos) {
-        console.log(pos);
-        return [];
+    _getPointsAtPosition(p, style) {
+        p = wToR(p.x, p.y, { center: this.center, scale: this.scale });
```
Comment (Jesus89) [inline src/core/dataframe.js]: `strokeWidth` is not taken into account here. I think we should add it to cover the whole point.

Comment (davidmanzanares) [reply src/core/dataframe.js]: Ups, good catch, I totally forgot about that.

Comment (davidmanzanares) [reply src/core/dataframe.js]: Solved!

File: test/unit/core/dataframe.test.js [hunk 3]
```diff
@@ -1,12 +1,61 @@
...
+            };
+            const feature2 = {
+                properties: {
+                    id: 2
+                }
+            };
+            const style = {
+                getWidth: () => ({
+                    eval: () => {
+                        return 0.5;
```
Comment (Jesus89) [inline test/unit/core/dataframe.test.js]: Use width 1 to allow the "click inside point but not over stroke" test

Comment (davidmanzanares) [reply test/unit/core/dataframe.test.js]: eh? :confused:

File: test/unit/core/dataframe.test.js [hunk 4]
```diff
@@ -1,12 +1,61 @@
...
+            it('should return an empty list when there are no points at the given position', () => {
+                expect(dataframe.getFeaturesAtPosition({ x: 0.5, y: 0.5 }, style)).toEqual([]);
+                expect(dataframe.getFeaturesAtPosition({ x: 5, y: 5 }, style)).toEqual([]);
+                expect(dataframe.getFeaturesAtPosition({ x: 1.0, y: 1.0 + 1.001 / 1024 }, style)).toEqual([]);
+            });
+            it('should return a list containing the features at the given position', () => {
+                expect(dataframe.getFeaturesAtPosition({ x: 0.0, y: 0.0 }, style)).toEqual([feature1]);
+                expect(dataframe.getFeaturesAtPosition({ x: 1.0, y: 1.0 }, style)).toEqual([feature2]);
+                expect(dataframe.getFeaturesAtPosition({ x: 1.0, y: 1.0 + 0.999 / 1024 }, style)).toEqual([feature2]);
+            });
```
Comment (Jesus89) [inline test/unit/core/dataframe.test.js]: Using width: 1.
- 0,0 inside point
- 0.6, 0.6 inside stroke

Comment (davidmanzanares) [reply test/unit/core/dataframe.test.js]: The inside stroke test is already being tested with  `x: 1.0, y: 1.0 + 0.999 / 1024 `

File: test/unit/core/dataframe.test.js [hunk 5]
```diff
@@ -1,12 +1,61 @@
...
+                }),
+                getStrokeWidth: () => ({
+                    eval: () => {
+                        return 0.5;
+                    }
+                })
+            };
+            dataframe.renderer = { _zoom: 1, gl: { canvas: { height: 1024 } } };
+            it('should return an empty list when there are no points at the given position', () => {
+                expect(dataframe.getFeaturesAtPosition({ x: 0.5, y: 0.5 }, style)).toEqual([]);
```
Comment (Jesus89) [inline test/unit/core/dataframe.test.js]: Using width: 1.
- 0,1.1 outside stroke
- 0,5 outside stroke

Comment (davidmanzanares) [reply test/unit/core/dataframe.test.js]: ???

File: src/core/dataframe.js [hunk 6]
```diff
@@ -61,26 +61,46 @@ export default class Dataframe {
...
+            const featureID = i / 2;
+            const center = {
+                x: points[i],
+                y: points[i + 1],
+            };
+            const f = {};
+            columnNames.forEach(name => {
+                f[name] = this.properties[name][featureID];
+            });
+            const scale = (styleWidth.eval(f) + styleStrokeWidth.eval(f)) * widthScale;
```
Comment (Jesus89) [inline src/core/dataframe.js]: I think we should use `(styleWidth.eval(f) + styleStrokeWidth.eval(f)) * widthScale * 0.5` here

Comment (davidmanzanares) [reply src/core/dataframe.js]: Why? have you manually tried the example? This matched the rendering in my tests

File: test/unit/core/dataframe.test.js (hunk 3 above)
Comment (Jesus89) [reply test/unit/core/dataframe.test.js]: With the same width and widthStroke the point is fully covered by the stroke and you can not tests the case when the mouse is inside the point but not over the stroke.

File: src/core/dataframe.js (hunk 6 above)
Comment (Jesus89) [reply src/core/dataframe.js]: Yep, so I am missing something. This is the model, right?
point

So may be the 0.5 is implicit in widthScale.
Could you spot the missing part (just for my clarification :wink:)

Comment (davidmanzanares) [reply src/core/dataframe.js]: I see your point, I think it's due to `1 / this.renderer.gl.canvas.height`. The vertex shader converts the coordinates to the [-1,1] range, therefore, a `2 / this.renderer.gl.canvas.height` is what we are doing at render time. Here the `2` is being canceled by the `0.5`.

File: test/unit/core/dataframe.test.js (hunk 3 above)
Comment (davidmanzanares) [reply test/unit/core/dataframe.test.js]: The inside stroke case is already being tested with x: 1.0, y: 1.0 + 0.999 / 1024

Comment (Jesus89) [reply test/unit/core/dataframe.test.js]: :+1:

File: test/unit/core/dataframe.test.js (hunk 4 above)
Comment (Jesus89) [reply test/unit/core/dataframe.test.js]: :+1:

File: src/core/dataframe.js (hunk 6 above)
Comment (Jesus89) [reply src/core/dataframe.js]: Ok. It would be nice to have a comment with this information in the code to avoid future misunderstandings.

**Comment by Jesus89** [comment]
Great! Now maths and code are aligned. Thanks :D

Merge time!

**Pull request merged**